### PR TITLE
storagemarket.GetLateMiners

### PR DESCRIPTION
**Pull request by shannonwells**
Support for #2939 

The fault monitor needs a list of storage miners that are late. Storage miner has the list of miners and can ask each actor for its lateness. Create an export in storage market that returns a pointer to a map of address strings with PoStStates, but only if that state is late.

### CODE REVIEW

File: actor/builtin/storagemarket/storagemarket.go [hunk 1]
```diff
@@ -92,17 +92,21 @@ var storageMarketExports = exec.Exports{
 		Params: []abi.Type{abi.BytesAmount, abi.PeerID},
 		Return: []abi.Type{abi.Address},
 	},
-	"updateStorage": &exec.FunctionSignature{
-		Params: []abi.Type{abi.BytesAmount},
-		Return: nil,
+	"getMiners": &exec.FunctionSignature{
```
Comment (shannonwells) [inline actor/builtin/storagemarket/storagemarket.go]: alphabetized

File: plumbing/api.go [hunk 2]
```diff
@@ -118,6 +118,9 @@ func (api *API) ActorLs(ctx context.Context) (<-chan state.GetAllActorsResult, e
 	return api.chain.LsActors(ctx)
 }
 
+// StorageMinerLs returns a list of storage miner addresses
+func (api *API) StorageMinerLs(ctx context.Context) {}
```
[the remark this answers is not in the document]
Comment (shannonwells) [reply plumbing/api.go]: Yes, I forgot about it. It's deleted locally and I hadn't pushed the changes.

File: actor/builtin/miner/miner.go [hunk 3]
```diff
@@ -516,6 +519,25 @@ func (ma *Actor) GetPoStState(ctx exec.VMContext) (*big.Int, uint8, error) {
 	return big.NewInt(result), 0, nil
 }
 
+// GetSlashable returns whether this miner is slashable. At this time this means only that
```
Comment (ZenGround0) [inline actor/builtin/miner/miner.go]: This name is a bit misleading.  The comment does a good job of explaining it but if I only saw the name I would for sure think this returns true when the miner actor can actually be slashed.  Because this returns true when there are sectors in the miner's set what about `GetProving`, `Proving`, `ActivelyProving` etc, something along those lines?

File: actor/builtin/miner/miner_test.go [hunk 4]
```diff
@@ -1133,6 +1146,30 @@ func TestMinerGetPoStState(t *testing.T) {
...
+		done := types.EmptyIntSet()
+		mal.requirePoSt(firstCommitBlockHeight+5, done)
+		// even though they are late, they have never submitted a post so
+		mal.assertSlashability(true, firstCommitBlockHeight+6)
+	})
+
+	t.Run("is reported as not slashable if no proving set", func(t *testing.T) {
+		mal := setupMinerActorLiason(t)
+		mal.assertSlashability(false, firstCommitBlockHeight)
+	})
```
Comment (ZenGround0) [inline actor/builtin/miner/miner_test.go]: It might also be good to test the edge case where a miner has a single sector in its proving set after its first commitSector but before its first submitPoSt

File: actor/builtin/storagemarket/storagemarket.go [hunk 5]
```diff
@@ -193,6 +197,53 @@ func (sma *Actor) UpdateStorage(vmctx exec.VMContext, delta *types.BytesAmount)
 	return 0, nil
 }
 
+func (sma *Actor) GetSlashableMiners(vmctx exec.VMContext) (*[]address.Address, uint8, error) {
```
Comment (ZenGround0) [inline actor/builtin/storagemarket/storagemarket.go]: Same comment about naming as above propagates to here.  Consider `GetProvingMiners`

File: actor/builtin/storagemarket/storagemarket_test.go [hunk 6]
```diff
@@ -234,3 +299,52 @@ func deriveMinerAddress(creator address.Address, nonce uint64) (address.Address,
 
 	return address.NewActorAddress(buf.Bytes())
 }
+
+// mustCreateStorageMiner creates a storage miner for the given state tree & storage map
+// and returns the address of the created miner.
+func mustCreateStorageMiner(t *testing.T, st state.Tree, vms vm.StorageMap, height uint64) address.Address {
```
Comment (ZenGround0) [inline actor/builtin/storagemarket/storagemarket_test.go]: Please consider building on top of existing work in testhelpers/core.go to consolidate our testing framework.  There are already `CreateTestMiner` and `CreateTestMinerWith`.  Neither is exactly what you want because you need to specify height but it should be straightforward to add this param to `CreateTestMinerWith` and then update callers.

File: testhelpers/core.go [hunk 7]
```diff
@@ -151,7 +151,8 @@ func CreateTestMinerWith(
 
 	result, err := ApplyTestMessage(stateTree, vms, msg, types.NewBlockHeight(0))
 	require.NoError(t, err)
-
+	require.NotNil(t, result)
+	require.NoError(t, result.ExecutionError)
```
Comment (ZenGround0) [inline testhelpers/core.go]: 🙏

File: actor/builtin/storagemarket/storagemarket.go (hunk 5 above)
Comment (shannonwells) [reply actor/builtin/storagemarket/storagemarket.go]: So I had thought about this -- I was allowing for the possibility that other factors besides nonzero proving set may be used to determine slashability.  I'm not married to the name though.

File: actor/builtin/miner/miner_test.go (hunk 4 above)
Comment (shannonwells) [reply actor/builtin/miner/miner_test.go]: I thought that was the last example. What's missing?

Comment (ZenGround0) [reply actor/builtin/miner/miner_test.go]: It's similar to the last example but you would run `mal.assertIsProving(true. firstCommitBlockHeight+1)` after the first `mal.requireCommit`

Comment (shannonwells) [reply actor/builtin/miner/miner_test.go]: ah. I was thinking about storagemarket_test where I'm doing that already, but I'll put an explicit one here.

File: actor/builtin/miner/miner.go [hunk 8]
```diff
@@ -516,6 +519,25 @@ func (ma *Actor) GetPoStState(ctx exec.VMContext) (*big.Int, uint8, error) {
...
+		return state.ProvingSet.Size() != 0, nil
+	})
+
+	if err != nil {
+		return false, errors.CodeError(err), err
+	}
+	result, ok := out.(bool)
+	if !ok {
+		return false, 1, errors.NewFaultErrorf("expected a bool, but got %T instead", out)
+	}
```
Comment (anorth) [inline actor/builtin/miner/miner.go]: FYI I added some other helpers to make read-only methods less verbose. See `GetProvingPeriod` for an example.

File: actor/builtin/storagemarket/storagemarket.go [hunk 9]
```diff
@@ -237,6 +288,19 @@ func (sma *Actor) GetProofsMode(vmctx exec.VMContext) (types.ProofsMode, uint8,
...
+	if err != nil {
+		return false, err
+	}
+
+	isProving, err := abi.Deserialize(msgResult[0], abi.Boolean)
+	if err != nil {
+		return false, err
+	}
+	return isProving.Val.(bool), nil
+}
```
Comment (anorth) [inline actor/builtin/storagemarket/storagemarket.go]: @acruikshank FWIW I think it's a bit silly that read-only methods that are not part of consensus have to go through this message-passing and ABI-deserialization junk. We're doing something wrong.

Comment (acruikshank) [reply actor/builtin/storagemarket/storagemarket.go]: (non-blocking)
I agree, but it's also kind of silly that state-manipulating methods that are part of consensus go through ABI-deserialization. Nothing about consensus requires this right now. It's all part of a long term plan to move the state machine into a separate VM that we haven't been able to prioritize.

There are advantages to keeping code that accesses actor-state within the actor. There are also advantages to limiting actors to message processing (actors not objects). There's no real advantage to serializing parameters and deserializing results (until we have a VM). We can simplify this interface in the near term.

One option would be a function that could generate a populated VMContext at the chain head. With this, you could call the actor method directly. See #3067 for an example of doing this in tests.

File: actor/builtin/miner/miner.go (hunk 8 above)
Comment (shannonwells) [reply actor/builtin/miner/miner.go]: ah, I didn't even see loadState.  Thanks for the pointer.

File: actor/builtin/storagemarket/storagemarket.go [hunk 10]
```diff
@@ -237,6 +288,19 @@ func (sma *Actor) GetProofsMode(vmctx exec.VMContext) (types.ProofsMode, uint8,
...
 
+func (sma *Actor) getMinerIsProving(vmctx exec.VMContext, minerAddr address.Address) (bool, error) {
+	msgResult, _, err := vmctx.Send(minerAddr, "isProving", types.ZeroAttoFIL, nil)
+	if err != nil {
+		return false, err
+	}
+
+	isProving, err := abi.Deserialize(msgResult[0], abi.Boolean)
+	if err != nil {
+		return false, err
```
Comment (acruikshank) [inline actor/builtin/storagemarket/storagemarket.go]: Blocking: This needs to be wrapped in a fault errror

File: actor/builtin/storagemarket/storagemarket.go [hunk 11]
```diff
@@ -193,6 +197,53 @@ func (sma *Actor) UpdateStorage(vmctx exec.VMContext, delta *types.BytesAmount)
...
+	ctx := context.Background()
+	ret, err := actor.WithState(vmctx, &state, func() (interface{}, error) {
+		lu, err := actor.LoadLookup(ctx, vmctx.Storage(), state.Miners)
+		if err != nil {
+			return nil, err
+		}
+
+		vals, err := lu.Values(ctx)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (shannonwells) [reply actor/builtin/storagemarket/storagemarket.go]: this functionality is now gone and GetPoStState already followed the error reporting protocol.

File: actor/builtin/storagemarket/storagemarket.go (hunk 10 above)
Comment (shannonwells) [reply actor/builtin/storagemarket/storagemarket.go]: GetProofsMode and GetTotalStorage don't wrap in fault errors -- should these be updated as well?

File: actor/builtin/storagemarket/storagemarket.go [hunk 12]
```diff
@@ -193,6 +198,56 @@ func (sma *Actor) UpdateStorage(vmctx exec.VMContext, delta *types.BytesAmount)
...
 }
 
+func (sma *Actor) GetLateMiners(vmctx exec.VMContext) (*map[string]uint64, uint8, error) {
+	if err := vmctx.Charge(actor.DefaultGasCost); err != nil {
+		return nil, exec.ErrInsufficientGas, errors.RevertErrorWrap(err, "Insufficient gas")
+	}
+	var state State
+	ctx := context.Background()
+
+	ret, err := actor.WithState(vmctx, &state, func() (interface{}, error) {
```
Comment (acruikshank) [inline actor/builtin/storagemarket/storagemarket.go]: This read-only method can be simplified a bit by using the new `actor.ReadState` method instead of `actor.WithState`

File: testhelpers/core.go [hunk 13]
```diff
@@ -113,11 +113,26 @@ func VMStorage() vm.StorageMap {
 	return vm.NewStorageMap(blockstore.NewBlockstore(datastore.NewMapDatastore()))
 }
 
-// CreateTestMiner creates a new test miner with the given peerID and miner
+// MustSign signs a given address with the provided mocksigner or panics if it
+// cannot.
+func MustSign(s types.MockSigner, msgs ...*types.Message) []*types.SignedMessage {
```
Comment (acruikshank) [inline testhelpers/core.go]: Why is this being added? It doesn't seem to be used in this PR.

Comment (shannonwells) [reply testhelpers/core.go]: I did not see that; it must have been pulled in when I did a rebase.
